File: src/types/machine.ts
export interface BiosSnapshot {
  manufacturer: string | null;
  version: string | null;
  releaseDate: string | null;
}

export interface MachineSystemSnapshot {
  hostname: string;
  username: string;
  operatingSystem: string | null;
  windowsVersion: string | null;
  architecture: string;
  manufacturer: string | null;
  model: string | null;
  cpu: string | null;
  ramBytes: number | null;
  bios: BiosSnapshot | null;
  error: string | null;
}

export interface VolumeSnapshot {
  unit: string;
  label: string | null;
  totalBytes: number | null;
  freeBytes: number | null;
  usedBytes: number | null;
}

export interface NetworkAdapterSnapshot {
  name: string;
  status: string | null;
  mac: string | null;
  ipv4: string[];
  ipv6: string[];
  gateways: string[];
  dnsServers: string[];
}

export interface PrinterSnapshot {
  name: string;
  isDefault: boolean;
  driver: string | null;
  port: string | null;
  server: string | null;
  status: string | null;
}

export interface SnapshotCollection<T> {
  items: T[];
  error: string | null;
}

export interface MachineSnapshot {
  /** Milliseconds since Unix epoch when collection began. */
  capturedAt: number;
  system: MachineSystemSnapshot;
  storage: SnapshotCollection<VolumeSnapshot>;
  network: SnapshotCollection<NetworkAdapterSnapshot>;
  printers: SnapshotCollection<PrinterSnapshot>;
}

/** Query shapes reserved for explicit, targeted collection in a later step. */
export interface ServiceQuery { names: string[] }
export interface ProcessQuery { name?: string; processId?: number }

File: src-tauri/src/models/machine.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct BiosSnapshot {
    pub manufacturer: Option<String>,
    pub version: Option<String>,
    pub release_date: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct MachineSystemSnapshot {
    pub hostname: String,
    pub username: String,
    pub operating_system: Option<String>,
    pub windows_version: Option<String>,
    pub architecture: String,
    pub manufacturer: Option<String>,
    pub model: Option<String>,
    pub cpu: Option<String>,
    pub ram_bytes: Option<u64>,
    pub bios: Option<BiosSnapshot>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct VolumeSnapshot {
    pub unit: String,
    pub label: Option<String>,
    pub total_bytes: Option<u64>,
    pub free_bytes: Option<u64>,
    pub used_bytes: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct NetworkAdapterSnapshot {
    pub name: String,
    pub status: Option<String>,
    pub mac: Option<String>,
    pub ipv4: Vec<String>,
    pub ipv6: Vec<String>,
    pub gateways: Vec<String>,
    pub dns_servers: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct PrinterSnapshot {
    pub name: String,
    pub is_default: bool,
    pub driver: Option<String>,
    pub port: Option<String>,
    pub server: Option<String>,
    pub status: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SnapshotCollection<T> {
    pub items: Vec<T>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct MachineSnapshot {
    /// Milliseconds since Unix epoch when collection began.
    pub captured_at: u64,
    pub system: MachineSystemSnapshot,
    pub storage: SnapshotCollection<VolumeSnapshot>,
    pub network: SnapshotCollection<NetworkAdapterSnapshot>,
    pub printers: SnapshotCollection<PrinterSnapshot>,
}

/// Request shape reserved for future targeted service lookups. No services are enumerated in a snapshot.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServiceQuery {
    pub names: Vec<String>,
}

/// Request shape reserved for future targeted process lookups. No processes are enumerated in a snapshot.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ProcessQuery {
    pub name: Option<String>,
    pub process_id: Option<u32>,
}
